File: translations/CJBench/J2P/DeepseekCommented/starcoder/Code/CountingInversions.py
import sys

# Function to count the number of inversions in an array
def countInversions(arr):
    return mergeSortAndCount(arr, 0, len(arr) - 1)

# Recursive method to sort the array and count inversions
def mergeSortAndCount(arr, left, right):
    if left >= right:
        return 0 # Base case: no inversions in a single element array
    mid = left + (right - left) / 2
    inversions = 0
    inversions += mergeSortAndCount(arr, left, mid) # Count inversions in the left half
    inversions += mergeSortAndCount(arr, mid + 1, right) # Count inversions in the right half
    inversions += mergeAndCount(arr, left, mid, right) # Count inversions during merge
    return inversions

# Method to merge two halves and count inversions during merge
def mergeAndCount(arr, left, mid, right):
    leftArr = arr[left:mid + 1]
    rightArr = arr[mid + 1:right + 1]
    i = 0
    j = 0
    k = left
    inversions = 0
    while i < len(leftArr) and j < len(rightArr):
        if leftArr[i] <= rightArr[j]:
            arr[k] = leftArr[i] # No inversion, copy from leftArr
            i += 1
        else:
            arr[k] = rightArr[j] # Inversion found, copy from rightArr
            inversions += len(leftArr) - i # Count inversions
            j += 1
        k += 1
    while i < len(leftArr):
        arr[k] = leftArr[i] # Copy remaining elements from leftArr
        i += 1
        k += 1
    while j < len(rightArr):
        arr[k] = rightArr[j] # Copy remaining elements from rightArr
        j += 1
        k += 1
    return inversions

# Main method to read input, compute inversions, and print the result
if __name__ == "__main__":
    n = int(raw_input().strip()) # Read the size of the array
    arr = [int(raw_input().strip()) for _ in xrange(n)]
    result = countInversions(arr) # Compute the number of inversions
    print "Number of inversions: " + str(result) # Print the result

File: data/CJBench/Java/Uncommented/CountingInversions.java

import java.util.Scanner;
public final class CountingInversions {
    private CountingInversions() {
    }
    public static int countInversions(int[] arr) {
        return mergeSortAndCount(arr, 0, arr.length - 1);
    }
    private static int mergeSortAndCount(int[] arr, int left, int right) {
        if (left >= right) {
            return 0;
        }
        int mid = left + (right - left) / 2;
        int inversions = 0;
        inversions += mergeSortAndCount(arr, left, mid);
        inversions += mergeSortAndCount(arr, mid + 1, right);
        inversions += mergeAndCount(arr, left, mid, right);
        return inversions;
    }
    private static int mergeAndCount(int[] arr, int left, int mid, int right) {
        int[] leftArr = new int[mid - left + 1];
        int[] rightArr = new int[right - mid];
        System.arraycopy(arr, left, leftArr, 0, mid - left + 1);
        System.arraycopy(arr, mid + 1, rightArr, 0, right - mid);
        int i = 0;
        int j = 0;
        int k = left;
        int inversions = 0;
        while (i < leftArr.length && j < rightArr.length) {
            if (leftArr[i] <= rightArr[j]) {
                arr[k++] = leftArr[i++];
            } else {
                arr[k++] = rightArr[j++];
                inversions += mid + 1 - left - i;
            }
        }
        while (i < leftArr.length) {
            arr[k++] = leftArr[i++];
        }
        while (j < rightArr.length) {
            arr[k++] = rightArr[j++];
        }
        return inversions;
    }
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int[] arr = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = scanner.nextInt();
        }
        int result = countInversions(arr);
        System.out.println("Number of inversions: " + result);
        scanner.close();
    }
}
